### PR TITLE
Add sine and cosine operators

### LINKED ISSUE #462: Add sin and cos operators
#365

### PR DESCRIPTION
**Pull request by soumyac1999**
Fixes #462.

Discussion items (@lockshaw):
1. Where and do we expose these ops in the Keras frontend?
2. Are there C++ and/or Python tests which need to be added corresponding to these new operations?

### CODE REVIEW

**Comment by lockshaw** [comment]
>     1. Where and do we expose these ops in the Keras frontend?

How are they being invoked in the current keras notebook you have?

>     2. Are there C++ and/or Python tests which need to be added corresponding to these new operations?

Unfortunately at the moment I don't think so. Feel free to add some if you want (we definitely want to move towards having more tests, but I don't think there's currently a systematic approach--see #429)

**Comment by lockshaw** [comment]
For now you could probably just access the core FF interface directly and we can discuss how to surface it later

File: python/flexflow/core/flexflow_cffi.py [hunk 1]
```diff
@@ -864,6 +882,39 @@ def exp(self, x, name=None):
     self.add_layer(OpType.EXP, name)
     return Tensor(handle, owner_op_type=OpType.EXP)
 
+  def sin(self, x, name=None):
+    """Sine function.
```
Comment (lockshaw) [inline python/flexflow/core/flexflow_cffi.py:886]:
```suggestion
    """Elementwise sine function.
```

File: python/flexflow/core/flexflow_cffi.py [hunk 2]
```diff
@@ -864,6 +882,39 @@ def exp(self, x, name=None):
...
+
+    :returns:  Tensor -- the output tensor.
+    """
+    c_name = get_c_name(name)
+    handle = ffc.flexflow_model_add_sin(self.handle, x.handle, c_name)
+    self.add_layer(OpType.SIN, name)
+    return Tensor(handle, owner_op_type=OpType.SIN)
+
+  def cos(self, x, name=None):
+    """Cosine function.
```
Comment (lockshaw) [inline python/flexflow/core/flexflow_cffi.py:902]:
```suggestion
    """Elementwise cosine function.
```

File: src/ops/element_unary.cu [hunk 3]
```diff
@@ -92,6 +92,12 @@ __global__ void elewise_unary_forward_kernel(
         out[i] = (T)(powf(in[i], scalar));
         break;
       }
+      case OP_SIN: {
+        out[i] = (T)sin((float)in[i]);
+      }
```
Comment (lockshaw) [inline src/ops/element_unary.cu:97]: You need `break` statements here

File: src/ops/element_unary.cpp [hunk 4]
```diff
@@ -92,6 +92,12 @@ __global__ void elewise_unary_forward_kernel(
         out[i] = (T)(powf(in[i], scalar));
         break;
       }
+      case OP_SIN: {
```
Comment (lockshaw) [inline src/ops/element_unary.cpp:95]: You need `break` statements here--C++ case by default falls through

**Review by lockshaw — CHANGES_REQUESTED** [review]

File: src/ops/element_unary.cpp [hunk 5]
```diff
@@ -92,6 +92,12 @@ __global__ void elewise_unary_forward_kernel(
...
+        out[i] = (T)sin((float)in[i]);
```
Comment (soumyac1999) [inline src/ops/element_unary.cpp:96]:
```suggestion
        out[i] = (T)sin((float)in[i]);
        break;
```

File: src/ops/element_unary.cpp [hunk 6]
```diff
@@ -92,6 +92,12 @@ __global__ void elewise_unary_forward_kernel(
...
+      }
+      case OP_COS: {
+        out[i] = (T)cos((float)in[i]);
```
Comment (soumyac1999) [inline src/ops/element_unary.cpp:99]:
```suggestion
        out[i] = (T)cos((float)in[i]);
        break;
```

File: src/ops/element_unary.cu [hunk 7]
```diff
@@ -92,6 +92,12 @@ __global__ void elewise_unary_forward_kernel(
         out[i] = (T)(powf(in[i], scalar));
         break;
       }
+      case OP_SIN: {
+        out[i] = (T)sin((float)in[i]);
```
Comment (soumyac1999) [inline src/ops/element_unary.cu:96]:
```suggestion
        out[i] = (T)sin((float)in[i]);
        break;
```

File: src/ops/element_unary.cu [hunk 8]
```diff
@@ -92,6 +92,12 @@ __global__ void elewise_unary_forward_kernel(
...
+      }
+      case OP_COS: {
+        out[i] = (T)cos((float)in[i]);
```
Comment (soumyac1999) [inline src/ops/element_unary.cu:99]:
```suggestion
        out[i] = (T)cos((float)in[i]);
        break;
```

**Review by lockshaw — APPROVED** [review]

**Pull request merged**